[FLINK-2808] Rework state abstraction and clean up task / operator internals

StephanEwen opened the pull request:
This pull request fixes many related/intermixed issues. It was hard to split this into individual issues.

### Crucial bug fixes

  - State snapshots for memory backed state previously copied a reference into the StateHandle, after which the streaming program continued. If the state was mutated prior to serialization by Akka, the mutated state was checkpointed, rather than the state at the point of drawing the snapshot.
  
  - Key/value state is checkpointed as a whole, rather than individually per key.
  
  - Memory-backed state now has a maximum size that is checked upon checkpointing. Exceeding that size fails the checkpoint. Before, too large state simply resulted in an oversized Akka frame that was dropped, silently letting the program run without ever completing a checkpoint.


### User-facing changes

  - The state backend is not only responsible for storing snapshots of the user state, but they also define how exactly the key/value state is represented in the first place. This allows us to plug in external key/value stores to store the Flink key/value state. Default implementations store the state in memory / files.
  
  - State backend offers additional methods to checkpoint directly into streams.
  
  - One can configure arbitrary default state backends via a factory interface that creates them from the TaskManager configuration.
  
  - Key/value state supports arbitrary types without extra checkpointer logic, but user needs to supply type of state (via class or TypeInformation)
  
  - Removed the `OperatorState` that is non-partitioned. The only type of state remaining through the `OperatorState` abstraction is partitioned key/value state in functions that are applied on a KeyedStream. Consequently, the `mapWithState()` and related methods are only available on the `KeyedStream`
  

### Internal cleanups

  - Checkpoint barriers are forwarded earlier, to reduce latency introduced by checkpoints.

  - Fewer in-memory copies when checkpointing to the file system state backend

  - The StreamingRuntimeContext is used purely for UDF interaction, not to hand over components to the operators. 

  - The infinite reduce and aggregations work properly on key/value state, rather than maintaining their own maps
  
  - made the OutputHandler (now called OperatorChain) type safe and simpler
  
  - made clear distinction between responsibilities of StreamTasks (input/output streams, setup of operator chain, checkpoint coordination) and operators (scope of one function and runtime context)
  
  - clean up checkpointing logic between operator (checkpoints generic key/value state) and UDF operators (checkpoint UDFs)
  
  - removed Configuration from operator open() method (was used in confusion with UDF open(Configuration())

gyfora commented:
Overall looks very good. :+1: 

I personally like having the non-partitioned operator states and I use them in my programs, so I would prefer keeping them.

gyfora commented:
So my argument is that the overhead of implementing the Checkpointed interface for simple counters , offsets and the like is simply too much (and annoying).

We can also introduce some annotations that the user can use to tag the state with. Then the only thing we need to make sure is so that these support some custom checkpointing logic. (like the state checkpointer interface)

Another thing we should consider is that the Checkpointed interface will never allow any lazy state access logic.

senorcarbone commented on flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -322,73 +311,84 @@ public String getName() {
+				// lock scope, they are an atomic operation regardless of the order in which they occur
+				// we immediately emit the checkpoint barriers, so the downstream operators can start
+				// their checkpoint work as soon as possible
+				operatorChain.broadcastCheckpointBarrier(checkpointId, timestamp);
```
That's a cool and valid addition to the technique. I guess we will some some throughput improvement :). We just need to make sure that the tasksToAcknowledge are always all the tasks (and not e.g. all sinks) for this to work properly which we do from what I remember.

StephanEwen replied:
Yes, we wait for an ack from all tasks. If sinks ack earlier than intermediate tasks, the checkpoint is still only complete once the intermediates acknowledge it.

senorcarbone commented:
on a related note, is anyone planning to complete snapshotting for cyclc graphs for the 0.10 release? Unfortunately I do not have time to work on it this week but it would be great to support it for 0.10 already

gyfora commented:
Other important thing:

I think it would be very useful to add some extra context information to the snapshot methods. Something that can be used to identify the operator that took the snapshot. Maybe a combination of task id + task index.

I am currently trying to implement a sql based backend, and I would really need to store this information so I can then work with the checkpoint data from the outside.

gyfora commented:
I can live without the non-partitioned OperatorStates, but the context info for the snapshots is very hard to work around I think.

ktzoumas commented:
@senorcarbone I think the cyclic graphs feature will be hard to get into 0.10

gyfora commented:
Also in the StateBackend interface a close method would come in handy, similarly to the initialize to close connections etc. in case of failures.

senorcarbone commented:
@ktzoumas that's understandable! We could consider it for the maintenance release maybe if there is enough time

aljoscha commented on flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphGenerator.java:
```diff
@@ -471,7 +475,11 @@ private StreamGraph generateInternal(List<StreamTransformation<?>> transformatio
+			streamGraph.setKey(transform.getId(), transform.getStateKeySelector(), keySerializer);
+		}
+		if (transform.getStateKeyType() != null) {
+			
```
empty if

StephanEwen commented:
Thanks for the fast and good feedback.

Concerning the removal of non-partitioned operator state:
  - As per prior discussion offline, I wanted to consolidate this to a core feature set, since it is hard to remove such features once they are released and endorsed
  - For simple states, let us add an annotation `@State` (with possible checkpointer/serializer `@State(checkpointer=new BloomFilterCheckpointer())`) which serves the same purpose and would be more lightweight even. I would put that on the list for the next release.

Multiple key/value states per operator: Will re-add that (simple effort), it is in fact very useful

Closing the state backend: Will add that, good idea

Concerning context info during checkpointing:
  - You should have access to the RuntimeContext when drawing the snapshot, which means you can access the task index. Let us also add the JobVertexID to the RuntimeContext and these values stay deterministic across restarts (state handles are stored under JobVertexID + task index).

  - I was thinking about giving access to the state backend in the `Checkpointed` interface methods, but let's make a design for this for the next release.

With addressing these issues, any objections against adding this?

gyfora commented:
As for the context, I did not mean to add it to the Checkpointed interface but to the state backend method calls. 

Other simpler solution for now would be to make the environment or the runtimecontext accessible from the backend, so it knows which task it does the checkpointing for. This is probably the cleanest solution for now until we figure out the exact requirements.

Otherwise :+1: from me.

mbalassi commented:
I think when drawing the snapshot with `Checkpointed` you are already inside a streaming `Function` or internal operator where one can easily access that information anyway. 
I think that is fine that way, I agree with @gyfora that access from the `StateBackend` is more important as that is potentially decoupled from drawing the snapshot itself.
As for other points listed by @StephanEwen they sound reasonable for me.

gyfora commented on flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/state/AbstractHeapKvState.java:
```diff
@@ -0,0 +1,145 @@
+	@Override
+	public V value() {
+		V value = state.get(currentKey);
+		return value != null ? value : defaultValue;
```
I think you should make a copy of the default value here. Otherwise you end up with the same objects for non primitive types.

StephanEwen replied:
Ah, just saw the comment. Probably right, I need to do this in an add-on commit.

StephanEwen commented:
Manually merged
